src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -297,73 +353,77 @@ namespace
+                assert_cast<ColumnArray &>(column).insert(field);
+                break;
+            }
+            case ValueType::vtTuple:
+            {
+                if (value.type() != Poco::MongoDB::ElementTraits<MongoDocument::Ptr>::TypeId)
+                    throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, expected Tuple, got type id = {} for column {}",
+                                    toString(value.type()), name);
 
-                    auto [parent_ptr, child_idx] = arrays.back();
-                    auto parent = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(*parent_ptr).value();
+                auto document = static_cast<const Poco::MongoDB::ConcreteElement<MongoDocument::Ptr> &>(value).value();
 
-                    if (child_idx >= parent->size())
-                    {
-                        arrays.pop_back();
+                ColumnTuple & column_tuple = assert_cast<ColumnTuple &>(column);
 
-                        if (dimension_idx == 0)
-                            break;
+                const auto * tuple_type = assert_cast<const DataTypeTuple *>(&data_type);
+                const Strings & names = tuple_type->getElementNames();
+                const DataTypes & nested_types = tuple_type->getElements();
 
-                        dimensions[dimension_idx].emplace_back(Array(dimensions[dimension_idx + 1].begin(), dimensions[dimension_idx + 1].end()));
-                        dimensions[dimension_idx + 1].clear();
+                for (const auto col_num : collections::range(0, column_tuple.tupleSize()))
+                {
+                    IColumn & nested_column = column_tuple.getColumn(col_num);
+                    const DataTypePtr & nested_type = nested_types[col_num];
 
+                    const auto & nested_name = names[col_num];
+                    auto nested_external_desc = ExternalResultDescription::getValueTypeWithNullable(nested_type);
+
+                    bool exists_in_current_document = document->exists(nested_name);
+                    if (!exists_in_current_document)
+                    {
+                        nested_column.insertDefault();
                         continue;
                     }
 
-                    Poco::MongoDB::Element::Ptr child = parent->get(static_cast<int>(child_idx));
-                    arrays.back().second += 1;
+                    const Poco::MongoDB::Element::Ptr nested_value = document->get(nested_name);
 
-                    if (child->type() == Poco::MongoDB::ElementTraits<MongoArray::Ptr>::TypeId)
-                    {
-                        arrays.emplace_back(child.get(), 0);
-                    }
-                    else if (child->type() == Poco::MongoDB::ElementTraits<Poco::MongoDB::NullValue>::TypeId)
+                    if (nested_value.isNull() || nested_value->type() == Poco::MongoDB::ElementTraits<Poco::MongoDB::NullValue>::TypeId)
```

Comment:
This logic with Null value and Nullable column below repeats logic from `generate()` method. Let's just do it in the begining of the `insertValue` function.